source/MaterialXGenShader/GenContext.h

```diff
@@ -1,86 +1,198 @@
-/// For BSDF nodes another use-case is to distinguish between evaluation in a direct lighting
-/// context and an indirect lighting context where different versions of the node's function
-/// is required.
-/// 
-class GenContext
-{
-  public:
-    /// Constructor, set the identifier for this context.
-    GenContext(int id) : _id(id) {}
-
-    /// Return the identifier for this context.
-    int id() const { return _id; }
+class GenContext;
+class GenUserData;
 
-    /// Add an extra argument to be used for the node function in this context.
-    void addArgument(const Argument& arg) { _arguments.push_back(arg); }
+using GenContextPtr = std::shared_ptr<GenContext>;
+using GenUserDataPtr = std::shared_ptr<GenUserData>;
+using ConstGenUserDataPtr = std::shared_ptr<const GenUserData>;
 
-    /// Return a list of extra argument to be used for the node function in this context.
-    const Arguments& getArguments() const { return _arguments; }
-
-    /// Set a function name suffix to be used for the node function in this context.
-    void setFunctionSuffix(const string& suffix) { _functionSuffix = suffix; }
-
-    /// Return the function name suffix to be used for the node function in this context.
-    const string& getFunctionSuffix() const { return _functionSuffix; }
+/// Base class for custom user data needed during shader generation.
+class GenUserData : public std::enable_shared_from_this<GenUserData>
+{
+  public:
+    virtual ~GenUserData() {}
+    
+    /// Return a shared pointer for this object.
+    GenUserDataPtr getSelf()
+    {
+        return shared_from_this();
+    }
+
+    /// Return a shared pointer for this object.
+    ConstGenUserDataPtr getSelf() const
+    {
+        return shared_from_this();
+    }
+
+    /// Return this object cast to a templated type.
+    template<class T> shared_ptr<T> asA()
+    {
+        return std::dynamic_pointer_cast<T>(getSelf());
+    }
+
+    /// Return this object cast to a templated type.
+    template<class T> shared_ptr<const T> asA() const
+    {
+        return std::dynamic_pointer_cast<const T>(getSelf());
+    }
+
+  protected:
+    GenUserData() {}
+};
 
-    /// Add an input suffix to be used for the node function in this context.
+/// A context class for shader generation.
+/// Used for thread local storage of data needed during shader generation.
+class GenContext
+{
+  public:
+    /// Constructor.
+    GenContext(ShaderGeneratorPtr sg);
+
+    /// Return shader generatior.
+    const ShaderGenerator& getShaderGenerator() const
+    {
+        return *_sg;
+    }
+    
+    /// Return shader generation options.
+    GenOptions& getOptions()
+    {
+        return _options;
+    }
+
+    /// Return shader generation options.
+    const GenOptions& getOptions() const
+    {
+        return _options;
+    }
+
+    /// Add to the search path used for finding source code.
+    void registerSourceCodeSearchPath(const string& path)
```

Comment:
It was needed on macOS/Linux because both FilePath and FileSearchPath can be constructed by a string. So compiler don't know which version of registerSourceCodeSearchPath to use.